### PR TITLE
Fix wrong estimation due to wrong encoding engine.

### PR DESCRIPTION
**Pull request by tkountis**
When using an estimator to constantly add new elements and not checking the estimate
then encoding engine is not upgraded to the Dense, which leads to two problems.
a. the default engine keeps consuming more and more resources and it is fundamentally very hungry and slow
b. upon calling estimate at the very end, the conversion gives very wrong results

The problem with the delayed conversion, between the encoders, is that the actual condition check, whether to upgrade, takes place during the `estimate()` call, rather than the `add()`.

This patch, moves the check to the `add()` instead.

Credit goes to @pveentjer for reporting the issue.

### CODE REVIEW

**Comment by pveentjer** [comment]
Can you also add a reproducer to the test code that failed before the change and passed after the change is applied?

I'll try it in with the simulator test.

**Comment by pveentjer** [comment]
I just verified this fix with Simulator and it seems to be running fine. The error is small and the throughput remains stable.

**Comment by pveentjer** [comment]
👍 

Thanks for the quick fix!

File: hazelcast/src/test/java/com/hazelcast/cardinality/CardinalityEstimatorStressTest.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import static org.junit.Assert.assertEquals;
+
+@RunWith(HazelcastSerialClassRunner.class)
+@Category(NightlyTest.class)
+public class CardinalityEstimatorStressTest extends HazelcastTestSupport {
+
```
Comment (tombujok) [inline hazelcast/src/test/java/com/hazelcast/cardinality/CardinalityEstimatorStressTest.java]: Nice test!

**Pull request merged**